File: frontend/src/types/inventory.ts
export type AppReconciliationState =
  | 'registered'
  | 'discovered_unregistered'
  | 'registry_missing_root'
  | 'registry_invalid_binding'
  | 'registry_invalid_config'
  | 'orphaned_runtime'
  | 'orphaned_nginx_site'

export type DomainReconciliationState =
  | 'managed'
  | 'discovered'
  | 'drifted'
  | 'disabled_in_nginx'
  | 'proxy_target_detected'
  | 'missing_document_root'

export type SslReconciliationState =
  | 'managed'
  | 'discovered'
  | 'expiring'
  | 'expired'
  | 'missing'
  | 'mismatch'
  | 'unbound'

export interface DiscoveredApplication {
  id: string
  name: string
  probable_type: string
  root_path: string
  git_path?: string | null
  environment?: string | null
  server_names: string[]
  nginx_site_path?: string | null
  systemd_unit?: string | null
  process_match?: string | null
  signals: string[]
  registered: boolean
  registered_id?: string | null
  reconciliation_state: AppReconciliationState
  runtime_status?: string | null
  registry_errors?: string[]
}

export interface NginxDiscoveredDomain {
  server_name: string
  site_path: string
  enabled: boolean
  ssl_enabled: boolean
  document_root?: string | null
  proxy_pass?: string | null
  certificate_path?: string | null
  in_database: boolean
  database_id?: string | null
  linked_app_id?: string | null
  reconciliation_state: DomainReconciliationState
}

export interface VpsInventorySummary {
  timestamp: string
  registered_apps: number
  discovered_apps: number
  unregistered_discovered_apps: number
  managed_domains: number
  discovered_domains: number
  domains_with_drift: number
  certificates_healthy: number
  certificates_expiring: number
  certificates_missing: number
  runtime_issues: number
}

export interface VpsInventoryResponse {
  summary: VpsInventorySummary
  applications: {
    registered_total: number
    discovered_total: number
    unregistered_discovered: number
    issues_count: number
    registered: DiscoveredApplication[]
    discovered: DiscoveredApplication[]
  }
  domains: {
    managed_total: number
    discovered_total: number
    drift_count: number
    managed: NginxDiscoveredDomain[]
    discovered: NginxDiscoveredDomain[]
  }
  ssl: {
    managed_total: number
    discovered_total: number
    expiring_count: number
    expired_count: number
    missing_count: number
    certificates: Array<{
      domain: string
      reconciliation_state: SslReconciliationState
      in_database: boolean
      nginx_bound: boolean
      days_remaining?: number | null
    }>
  }
}

export type TerminalScope = 'ops' | 'hosting' | 'app'

File: backend/app/schemas/inventory.py
"""VPS discovery and reconciliation schemas."""

from datetime import datetime
from enum import StrEnum

from pydantic import Field

from app.schemas.common import SchemaBase
from app.schemas.health import HealthStatus


class AppReconciliationState(StrEnum):
    REGISTERED = "registered"
    DISCOVERED_UNREGISTERED = "discovered_unregistered"
    REGISTRY_MISSING_ROOT = "registry_missing_root"
    REGISTRY_INVALID_BINDING = "registry_invalid_binding"
    REGISTRY_INVALID_CONFIG = "registry_invalid_config"
    ORPHANED_RUNTIME = "orphaned_runtime"
    ORPHANED_NGINX_SITE = "orphaned_nginx_site"


class DomainReconciliationState(StrEnum):
    MANAGED = "managed"
    DISCOVERED = "discovered"
    DRIFTED = "drifted"
    DISABLED_IN_NGINX = "disabled_in_nginx"
    PROXY_TARGET_DETECTED = "proxy_target_detected"
    MISSING_DOCUMENT_ROOT = "missing_document_root"


class SslReconciliationState(StrEnum):
    MANAGED = "managed"
    DISCOVERED = "discovered"
    EXPIRING = "expiring"
    EXPIRED = "expired"
    MISSING = "missing"
    MISMATCH = "mismatch"
    UNBOUND = "unbound"


class DiscoveredApplicationSchema(SchemaBase):
    id: str
    name: str
    probable_type: str
    root_path: str
    git_path: str | None = None
    environment: str | None = None
    server_names: list[str] = Field(default_factory=list)
    nginx_site_path: str | None = None
    systemd_unit: str | None = None
    process_match: str | None = None
    signals: list[str] = Field(default_factory=list)
    registered: bool = False
    registered_id: str | None = None
    reconciliation_state: AppReconciliationState
    runtime_status: str | None = None
    registry_errors: list[str] = Field(default_factory=list)


class ApplicationInventorySchema(SchemaBase):
    timestamp: datetime
    registered_total: int
    discovered_total: int
    unregistered_discovered: int
    issues_count: int
    registered: list[DiscoveredApplicationSchema]
    discovered: list[DiscoveredApplicationSchema]


class NginxDiscoveredDomainSchema(SchemaBase):
    server_name: str
    site_path: str
    enabled: bool
    ssl_enabled: bool
    document_root: str | None = None
    proxy_pass: str | None = None
    certificate_path: str | None = None
    in_database: bool = False
    database_id: str | None = None
    linked_app_id: str | None = None
    reconciliation_state: DomainReconciliationState


class DomainInventorySchema(SchemaBase):
    timestamp: datetime
    managed_total: int
    discovered_total: int
    drift_count: int
    managed: list[NginxDiscoveredDomainSchema]
    discovered: list[NginxDiscoveredDomainSchema]


class DiscoveredCertificateSchema(SchemaBase):
    domain: str
    certificate_path: str | None = None
    issuer: str | None = None
    valid_until: datetime | None = None
    days_remaining: int | None = None
    status: HealthStatus | None = None
    sans: list[str] = Field(default_factory=list)
    in_database: bool = False
    nginx_bound: bool = False
    reconciliation_state: SslReconciliationState


class SslInventorySchema(SchemaBase):
    timestamp: datetime
    managed_total: int
    discovered_total: int
    expiring_count: int
    expired_count: int
    missing_count: int
    certificates: list[DiscoveredCertificateSchema]


class VpsInventorySummarySchema(SchemaBase):
    timestamp: datetime
    registered_apps: int
    discovered_apps: int
    unregistered_discovered_apps: int
    managed_domains: int
    discovered_domains: int
    domains_with_drift: int
    certificates_healthy: int
    certificates_expiring: int
    certificates_missing: int
    runtime_issues: int


class VpsInventoryResponse(SchemaBase):
    summary: VpsInventorySummarySchema
    applications: ApplicationInventorySchema
    domains: DomainInventorySchema
    ssl: SslInventorySchema
